pkg/vere/io/lick.c

```diff
@@ -0,0 +1,645 @@
+
+    u3_auto_peer(
+      u3_auto_plan(&lic_u->car_u,
+                   u3_ovum_init(0, c3__l, wir, cad)),
+      0, 0, 0);
+  }
+
+  u3_newt_moat_stop((u3_moat*)&can_u->mor_u, _lick_moat_free);
+}
+
+/* _lick_moor_bail(): error callback for u3_moor.
+*/
+static void
+_lick_moor_bail(void* ptr_v, ssize_t err_i, const c3_c* err_c)
+{
+  u3_chan*  can_u = (u3_chan*)ptr_v;
+
+  if ( err_i != UV_EOF ) {
+    u3l_log("lick: moor bail %zd %s", err_i, err_c);
+    if ( _(can_u->liv_o) ) {
+      _lick_send_noun(can_u, u3nq(0, c3__bail, u3i_word(err_i),
+                      u3i_string(err_c)));
+      can_u->liv_o = c3n;
+    }
+  }
+  _lick_close_chan(can_u);
+}
+
+/* _lick_sock_cb(): socket connection callback.
+*/
+static void
+_lick_sock_cb(uv_stream_t* sem_u, c3_i tas_i)
+{
+  u3_shan*  san_u = (u3_shan*)sem_u;
+  u3_port*  gen_u = san_u->gen_u;
+  u3_chan*  can_u;
+  c3_i      err_i;
+
+  can_u = c3_calloc(sizeof(u3_chan));
+  can_u->mor_u.ptr_v = can_u;
+  can_u->mor_u.pok_f = _lick_moor_poke;
+  can_u->mor_u.bal_f = _lick_moor_bail;
+  can_u->coq_l = san_u->nex_l++;
+  can_u->san_u = san_u;
+  err_i = uv_timer_init(u3L, &can_u->mor_u.tim_u);
+  u3_assert(!err_i);
+  err_i = uv_pipe_init(u3L, &can_u->mor_u.pyp_u, 0);
+  u3_assert(!err_i);
+  err_i = uv_accept(sem_u, (uv_stream_t*)&can_u->mor_u.pyp_u);
+  u3_assert(!err_i);
+  u3_newt_read((u3_moat*)&can_u->mor_u);
+  can_u->mor_u.nex_u = (u3_moor*)san_u->can_u;
+  san_u->can_u = can_u;
+  gen_u->con_o = c3y;
+
+  // only send on first connection
+  if ( NULL == can_u->mor_u.nex_u ) {
+    u3_noun   dev, dat, wir, cad, mar;
+    wir = u3nc(c3__lick, u3_nul);
+    dev = _lick_string_to_path(gen_u->nam_c+1);
+    mar = u3i_string("connect");
+    dat = u3_nul;
+
+    cad = u3nq(c3__soak, dev, mar, dat);
+    u3_auto_peer(
+      u3_auto_plan(&gen_u->lic_u->car_u, u3_ovum_init(0, c3__l, wir, cad)),
+      0, 0, 0);
+  }
+}
+
+/* _lick_close_sock():  close an port's socket
+*/
+static void
+_lick_close_sock(u3_shan* san_u)
+{
+  c3_c* pax_c = u3_Host.dir_c;
+  c3_w len_w = strlen(pax_c) + 2 + sizeof(URB_DEV_PATH) + strlen(san_u->gen_u->nam_c);
+  c3_c* paf_c = c3_malloc(len_w);
+  c3_i  wit_i;
+  wit_i = snprintf(paf_c, len_w, "%s/%s/%s", pax_c, URB_DEV_PATH, san_u->gen_u->nam_c);
+
+  u3_assert(wit_i > 0);
+  u3_assert(len_w == (c3_w)wit_i + 1);
+
+  if ( 0 != unlink(paf_c) ) {
+    if ( ENOENT != errno ) {
+      u3l_log("lick: failed to unlink socket: %s", uv_strerror(errno));
+    }
+  }
+
+  uv_close((uv_handle_t*)&san_u->pyp_u, _lick_close_cb);
+  c3_free(paf_c);
+}
+
+/* _lick_mkdirp(): recursive mkdir of dirname of pax_c.
+*/
+static void
+_lick_mkdirp(c3_c* por_c)
+{
+  c3_c pax_c[2048];
+
+  strncpy(pax_c, por_c, sizeof(pax_c));
+  
+  c3_c* fas_c = strchr(pax_c + 1, '/');
+
+  while ( fas_c ) {
+    *fas_c = 0;
+    if ( 0 != mkdir(pax_c, 0777) && EEXIST != errno ) {
+      u3l_log("lick: mkdir %s: %s", pax_c, strerror(errno));
+      u3m_bail(c3__fail);
```

Comment:
This is a fatal error, bailing on the home road will take down the king process. Better to return an error code and handle that in `_lick_init_sock()` along with the other i/o errors that are possible there.

Reply:
Done